tests/gen/gen_predefined.go

```diff
@@ -0,0 +1,178 @@
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+// GeneratePredefined generates documents from a predefined list
+// of docs that might include nested docs.
+func GeneratePredefined(defs []client.CollectionDefinition, docsList DocsList) ([]GeneratedDoc, error) {
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, col := range defs {
+		typeDefs[col.Description.Name] = col
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.generateRelatedDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+type docGenerator struct {
+	types map[string]client.CollectionDefinition
+}
+
+// toRequestedDoc removes the fields that are not in the SDL of the collection.
+func toRequestedDoc(doc map[string]any, typeDef *client.CollectionDefinition) map[string]any {
+	result := make(map[string]any)
+	for _, field := range typeDef.Schema.Fields {
+		if field.IsRelation() || field.Name == request.KeyFieldName {
+			continue
+		}
+		result[field.Name] = doc[field.Name]
+	}
+	for name, val := range doc {
+		if strings.HasSuffix(name, request.RelatedObjectID) {
+			result[name] = val
+		}
+	}
+	return result
+}
+
+// generatePrimary generates primary docs for the given secondary doc.
+func (this *docGenerator) generatePrimary(
+	docMap map[string]any,
+	typeDef *client.CollectionDefinition,
+) (map[string]any, []GeneratedDoc, error) {
+	result := []GeneratedDoc{}
+	requested := toRequestedDoc(docMap, typeDef)
+	for _, field := range typeDef.Schema.Fields {
+		if field.IsRelation() {
+			if _, hasProp := docMap[field.Name]; hasProp {
+				if field.IsPrimaryRelation() {
+					subType := this.types[field.Schema]
+					subDocMap := toRequestedDoc(docMap[field.Name].(map[string]any), &subType)
+					subDoc, err := client.NewDocFromMap(subDocMap)
+					if err != nil {
+						return nil, nil, NewErrFailedToGenerateDoc(err)
+					}
+					requested[field.Name+request.RelatedObjectID] = subDoc.Key().String()
+					result = append(result, GeneratedDoc{Col: &subType, Doc: subDoc})
+				}
+			}
+		}
+	}
+	return requested, result, nil
+}
+
+// generateRelatedDocs generates related docs (primary and secondary) for the given doc.
+func (this *docGenerator) generateRelatedDocs(docMap map[string]any, typeName string) ([]GeneratedDoc, error) {
+	typeDef := this.types[typeName]
+
+	requested, result, err := this.generatePrimary(docMap, &typeDef)
+	if err != nil {
+		return nil, err
+	}
+	doc, err := client.NewDocFromMap(requested)
+	if err != nil {
+		return nil, NewErrFailedToGenerateDoc(err)
+	}
+
+	result = append(result, GeneratedDoc{Col: &typeDef, Doc: doc})
+
+	var docKey string
+	for _, field := range typeDef.Schema.Fields {
```

Comment:
question: Why do you loop through the fields outside of `generateSecondaryDocs`, but inside of `generatePrimary`?

Reply:
I don't understand your question. What do you mean "inside"/"outside"?
There are loops inside of both

Reply:
`generateRelatedDocs` contains:
```
// no loop
generatePrimary()
...

//loop
for _, field := range typeDef.Schema.Fields {
   if field.IsRelation() {
   ...
   generateSecondaryDocs
```

But `generatePrimary` contains:
```
for _, field := range typeDef.Schema.Fields {
   if field.IsRelation() {
   ...
   foo
```
Is there a reason for this inconsistency or is it just how the code happened to be written?

Reply:
restructured